### ISSUE TITLE
Publisher apps reporting "Timeout was thrown when waiting for channel to attach"

### ISSUE THREAD

**Issue by QuintinWillison** [ably/ably-asset-tracking-android#859]
We have a customer who is reporting seeing [this exception](https://github.com/ably/ably-asset-tracking-android/blob/a75fa17dcba987d86e726c7f937f740cb5b6f8c5/common/src/main/java/com/ably/tracking/common/Ably.kt#L890) quite a lot via their Crashlytics reports:

```
Non-fatal Exception: s3.c: Timeout was thrown when waiting for channel to attach
       at com.ably.tracking.common.DefaultAbly.waitForChannelReconnection(Ably.kt:890)
       at com.ably.tracking.common.DefaultAbly.access$waitForChannelReconnection(Ably.kt:247)
       at com.ably.tracking.common.DefaultAbly$waitForChannelReconnection$1.invokeSuspend(Ably.kt)
       at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
       at kotlinx.coroutines.internal.ScopeCoroutine.afterResume(Scopes.kt:33)
       at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:102)
       at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
       at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
       at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
       at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
       at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
```

This is because of a 10 second timeout [baked into](https://github.com/ably/ably-asset-tracking-android/blob/a75fa17dcba987d86e726c7f937f740cb5b6f8c5/common/src/main/java/com/ably/tracking/common/Ably.kt#L872) the AAT implementation. The value of this timeout is likely arbitrary (confirmed by @KacperKluka) and could be a sign that we're fighting or otherwise overriding behaviours of the underlying, core SDK. 🤔 

From [this Slack thread (internal)](https://ably-real-time.slack.com/archives/C01EPJENRD0/p1671616911203099) where we've initially discussed this report...

@paddybyers said:

> Attaching to a channel can fail, either for reasons that aren't going to change, or for reasons that are transient. Adding a trackable, and getting it to the state where it is working (ie also attached) is therefore also subject to failure, where a retry might succeed.
> So there are two questions:
> 1. what developer experience do we want to expose here? Should we always pass failures (including failures for transient reasons) back to the caller and abort the attempted operation? Or should AAT keep retrying all the time that there's reason to believe that a retry might succeed? What's the currently implemented API, and do we think this is the correct API?
> 2. What should AAT do in respect of interacting with the underlying Ably client? If it simply times out, then it has no chance of passing a meaningful error back to the caller, or taking meaningful action itself. It should at least be waiting for the underlying library to complete the operation (successfully or otherwise) because then it can know whether or not the operation is retriable, and it can pass meaningful information back to the caller.
> So IMO the question isn't whether 10s is long enough or not. Simply timing out, and not waiting to find out from the underlying library what happened, is just wrong IMO

@davyskiba added some further analysis:

> I think this is not causing a crash, the provided logs start with “Non-fatal Exception”.
> In the code, it seems we’re never just throwing that exception outside - it is caught and returned via the callback or wrapped in a result.
> I don’t mean that this is not an issue, just wanted to point out that it looks like it is not an app crash.
> The reconnect method is called every time a publisher attempts to send a location. Even twice if the connection keeps on failing.
> The occurrence numbers might be quite inflated 🤔

**Comment by lawrence-forooghian** [comment]
I started looking into this whilst on some downtime from another issue, but I've moved back to that issue now so I'm going to unassign this one from me so somebody else can pick it up.

My plan was to document the following:

- the current error handling, recovery, and timeout behaviour in AAT Android
- the error handling, recovery, and timeout behaviour of the Ably client libraries

My intention was that we would then define an error handling / recovery behaviour for AAT that took advantage of everything provided to us by the Ably client libraries, and that we only override or preempt the client library behaviour with good cause.

So far, all I've done is to start documenting the first bullet point above (AAT behaviour). I'm posting here what I've written so far in case it's useful to somebody else.

## Asset Tracking Publisher SDK (Android)

As of Android repo `63c5295`.

### `Ably.kt`

It seems like all exceptions thrown by this file are wrapped in a `ConnectionException` - I don’t think it lets anything leak out.

- `waitForChannelReconnection(channel, timeout)`
    - this one just waits for channel to get into attached state, with specified timeout
    - if timeout elapses, then it throws `ConnectionException`
- which is called by `retryChannelOperationIfConnectionResumeFails`
    - if channel is suspended it first waits for it to become attached (using the above)
    - then it performs the operation on the channel
    - and if that operation throws a `message == "Connection resume failed" && code == 50000 && statusCode == 500` exception then it retries the above two steps
    - if the retry fails then it rethrows
- which is called by `disconnectChannel` / `private sendMessage` / `public updatePresenceData`

Auth token issue retrying:

- `private enterChannelPresence`
	- tries to enter presence on the channel
	- if it catches a `ConnectionException` with `errorInformation.code == AUTH_TOKEN_CAPABILITY_ERROR_CODE` (40160), then it calls its own `renewAuthSuspending` – which in turn calls `ably.auth.renewAuth` – and then attaches to the channel and tries to enter presence on the channel
	- if the retry fails then it rethrows

I remember that the idea of this one was for a use case where a capability for a given channel is only granted to a user once they are assigned trackable that corresponds to that channel, but that this capability might not yet be reflected in their latest auth token.

Closing channel:

- `private tryDisconnectChannel`	
	- calls `disconnectChannel` and ignores any thrown exceptions

Connecting:

- `private connectSuspending(timeout)`
	- waits for connection to become `CONNECTED`, calling `connect()` if necessary
	- if starts in or enters `FAILED` state, it throws a `ConnectionException`
	- if it doesn’t enter into `CONNECTED` state within the timeout then it throws a `ConnectionException`

### The rest of the SDK

What happens to the exceptions that come out of `Ably.kt`? Are there other retries going on?

Let's look for `"catch"`, `"timeout"`, `"retry"` in the rest of the publisher codebase.

`catch`:

- `ConnectionCreatedWorker` catches a `ConnectionException`

`timeout`:

- `Publisher#stop(timeoutInMilliseconds:)`
	- This gets passed to `StopWorker`.
	
`retry`:

- TODO there seem to be a fair few of these in the codebase, not gone through them yet

**Comment by deanna-lad** [comment]
The link to the corresponding Jira issue is https://ably.atlassian.net/browse/SDK-3194

**Comment by ikbalkaya** [comment]
This issue might be a result of https://github.com/ably/ably-java/issues/885 where channel fails to reattach because of attach/detach race. 
I'm wondering if we have any timings of these events happening and see if times matches logs described https://github.com/ably/ably-asset-tracking-android/issues/863 

I think it would be good to filter that possibility

**Comment by lawrence-forooghian** [comment]
If we remove the channel attach timeout, we need to consider #912.

**Comment by lawrence-forooghian** [comment]
I'm going to look into removing this AAT arbitrary timeout.

**Comment by davyskiba** [comment]
Is this getting resolved by [#965](https://github.com/ably/ably-asset-tracking-android/pull/965)?
